File: thrift/lib/py/protocol/TSimpleJSONProtocol.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements. See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership. The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License. You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.
#

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

from base64 import b64encode, b64decode

from .TProtocol import *

JSON_OBJECT_START = b'{'
JSON_OBJECT_END = b'}'
JSON_ARRAY_START = b'['
JSON_ARRAY_END = b']'
JSON_NEW_LINE = b'\n'
JSON_PAIR_SEPARATOR = b':'
JSON_ELEM_SEPARATOR = b','
JSON_BACKSLASH = b'\\'
JSON_STRING_DELIMITER = b'"'
JSON_ZERO_CHAR = b'0'
JSON_TAB = b"  "

JSON_ESCAPE_CHAR = b'u'
JSON_ESCAPE_PREFIX = b"\\u00"

THRIFT_VERSION_1 = 1

THRIFT_NAN = b"NaN"
THRIFT_INFINITY = b"Infinity"
THRIFT_NEGATIVE_INFINITY = b"-Infinity"

JSON_CHAR_TABLE = [ \
#   0   1    2   3   4   5   6   7    8    9    A   B    C    D   E   F
    0,  0,   0,  0,  0,  0,  0,  0,b'b',b't',b'n',  0,b'f',b'r',  0,  0, \
    0,  0,   0,  0,  0,  0,  0,  0,   0,   0,   0,  0,   0,   0,  0,  0, \
    1,  1,b'"',  1,  1,  1,  1,  1,   1,   1,   1,  1,   1,   1,  1,  1, \
]

ESCAPE_CHARS = b"\"\\bfnrt"
ESCAPE_CHAR_VALS = [b'"', b'\\', b'\b', b'\f', b'\n', b'\r', b'\t']


def hexChar(x):
    x &= 0x0f
    return hex(x)[2:]


def hexVal(ch):
    if ch >= '0' and ch <= '9':
        return int(ch) - int('0')
    elif ch >= 'a' and ch <= 'f':
        return int(ch) - int('a') + 10
    raise TProtocolException(TProtocolException.INVALID_DATA,
                             "Unexpected hex value")


class TJSONContext:
    def __init__(self, indentLevel=0):
        self.indentLevel = indentLevel
        return

    def write(self, trans):
        return

    def escapeNum(self):
        return False

    def writeNewLine(self, trans):
        trans.write(JSON_NEW_LINE)
        self.indent(trans)

    def indent(self, trans):
        for i in range(self.indentLevel):
            trans.write(JSON_TAB)


class TJSONPairContext(TJSONContext):
    def __init__(self, indentLevel=0, isMapPair=False):
        self.indentLevel = indentLevel
        self.first = True
        self.colon = True
        self.isMapPair = isMapPair

    def write(self, trans):
        if self.first:
            self.first = False
            self.colon = True
        else:
            if self.colon:
                trans.write(JSON_PAIR_SEPARATOR + b" ")
            else:
                trans.write(JSON_ELEM_SEPARATOR)
                if self.isMapPair:
                    self.writeNewLine(trans)
            self.colon = not self.colon

    def escapeNum(self):
        return self.colon


class TJSONListContext(TJSONContext):
    def __init__(self, indentLevel=0):
        self.indentLevel = indentLevel
        self.first = True

    def write(self, trans):
        if self.first:
            self.first = False
        else:
            trans.write(JSON_ELEM_SEPARATOR)
            self.writeNewLine(trans)


class TSimpleJSONProtocol(TProtocolBase):
    """
    JSON protocol implementation for Thrift. This protocol is write-only, and
    produces a simple output format that conforms to the JSON standard.
    """

    def __init__(self, trans):
        TProtocolBase.__init__(self, trans)
        self.contexts = [TJSONContext()]  # Used as stack for contexts.
        self.context = TJSONContext()

    def pushContext(self, newContext):
        self.contexts.append(self.context)
        self.context = newContext

    def popContext(self):
        if len(self.contexts) > 0:
            self.context = self.contexts.pop()

    # Writing functions.

    def writeJSONEscapeChar(self, ch):
        self.trans.write(JSON_ESCAPE_PREFIX)
        self.trans.write(hexChar(ch >> 4))
        self.trans.write(hexChar(ch))

    def writeJSONChar(self, ch):
        if ord(ch) >= 0x30:
            if ch == JSON_BACKSLASH:  # Only special character >= 0x30 is '\'.
                self.trans.write(JSON_BACKSLASH)
                self.trans.write(JSON_BACKSLASH)
            else:
                self.trans.write(ch)
        else:
            outCh = JSON_CHAR_TABLE[ord(ch)]
            if outCh == 1:
                self.trans.write(ch)
            elif outCh > 1:
                self.trans.write(JSON_BACKSLASH)
                self.trans.write(outCh)
            else:
                self.writeJSONEscapeChar(ord(ch))

    def writeJSONString(self, outStr):
        self.context.write(self.trans)
        self.trans.write(JSON_STRING_DELIMITER)
        for ch in outStr:
            self.writeJSONChar(ch)
        self.trans.write(JSON_STRING_DELIMITER)

    def writeJSONBase64(self, outStr):
        self.context.write(self.trans)
        self.trans.write(JSON_STRING_DELIMITER)
        b64Str = b64encode(outStr)
        self.trans.write(b64Str)
        self.trans.write(JSON_STRING_DELIMITER)

    def writeJSONInteger(self, num):
        self.context.write(self.trans)
        escapeNum = self.context.escapeNum()
        numStr = str(num)
        if escapeNum:
            self.trans.write(JSON_STRING_DELIMITER)
        self.trans.write(numStr)
        if escapeNum:
            self.trans.write(JSON_STRING_DELIMITER)

    def writeJSONBool(self, boolVal):
        self.context.write(self.trans)
        if self.context.escapeNum():
            self.trans.write(JSON_STRING_DELIMITER)
        if boolVal:
            self.trans.write(b"true")
        else:
            self.trans.write(b"false")
        if self.context.escapeNum():
            self.trans.write(JSON_STRING_DELIMITER)

    def writeJSONDouble(self, num):
        self.context.write(self.trans)
        numStr = str(num)
        special = False

        if numStr == "nan":
            numStr = THRIFT_NAN
            special = True
        elif numStr == "inf":
            numStr = THRIFT_INFINITY
            special = True
        elif numStr == "-inf":
            numStr = THRIFT_NEGATIVE_INFINITY
            special = True

        escapeNum = special or self.context.escapeNum()
        if escapeNum:
            self.trans.write(JSON_STRING_DELIMITER)
        self.trans.write(numStr)
        if escapeNum:
            self.trans.write(JSON_STRING_DELIMITER)

    def writeJSONObjectStart(self):
        self.context.write(self.trans)
        self.trans.write(JSON_OBJECT_START)
        self.pushContext(TJSONPairContext(len(self.contexts)))

    def writeJSONObjectEnd(self):
        self.popContext()
        self.context.writeNewLine(self.trans)
        self.trans.write(JSON_OBJECT_END)

    def writeJSONArrayStart(self):
        self.context.write(self.trans)
        self.trans.write(JSON_ARRAY_START)
        self.pushContext(TJSONListContext(len(self.contexts)))

    def writeJSONArrayEnd(self):
        self.popContext()
        self.context.writeNewLine(self.trans)
        self.trans.write(JSON_ARRAY_END)

    def writeJSONMapStart(self):
        self.context.write(self.trans)
        self.trans.write(JSON_OBJECT_START)
        self.pushContext(TJSONListContext(len(self.contexts)))

    def writeJSONMapEnd(self):
        self.popContext()
        self.context.writeNewLine(self.trans)
        self.trans.write(JSON_OBJECT_END)

    def writeMessageBegin(self, name, messageType, seqId):
        self.writeJSONArrayStart()
        self.pushContext(TJSONListContext(len(self.contexts)))
        self.writeJSONInteger(THRIFT_VERSION_1)
        self.writeJSONString(name)
        self.writeJSONInteger(messageType)
        self.writeJSONInteger(seqId)

    def writeMessageEnd(self):
        self.popContext()
        self.writeJSONArrayEnd()

    def writeStructBegin(self, name):
        self.writeJSONObjectStart()

    def writeStructEnd(self):
        self.writeJSONObjectEnd()

    def writeFieldBegin(self, name, fieldType, fieldId):
        self.context.write(self.trans)
        self.popContext()
        self.pushContext(TJSONPairContext(len(self.contexts)))
        self.context.writeNewLine(self.trans)
        self.writeJSONString(name)

    def writeFieldEnd(self):
        return

    def writeFieldStop(self):
        return

    def writeMapBegin(self, keyType, valType, size):
        self.writeJSONMapStart()
        self.context.writeNewLine(self.trans)
        self.pushContext(TJSONPairContext(len(self.contexts) - 1, True))

    def writeMapEnd(self):
        self.popContext()
        self.writeJSONMapEnd()

    def writeListBegin(self, elemType, size):
        self.writeJSONArrayStart()
        self.context.writeNewLine(self.trans)

    def writeListEnd(self):
        self.writeJSONArrayEnd()

    def writeSetBegin(self, elemType, size):
        self.writeJSONArrayStart()
        self.context.writeNewLine(self.trans)

    def writeSetEnd(self):
        self.writeJSONArrayEnd()

    def writeBool(self, val):
        self.writeJSONBool(val)

    def writeByte(self, byte):
        self.writeJSONInteger(byte)

    def writeI16(self, i16):
        self.writeJSONInteger(i16)

    def writeI32(self, i32):
        self.writeJSONInteger(i32)

    def writeI64(self, i64):
        self.writeJSONInteger(i64)

    def writeDouble(self, d):
        self.writeJSONDouble(d)

    def writeFloat(self, f):
        self.writeJSONDouble(f)

    def writeString(self, outStr):
        self.writeJSONString(outStr)

    def writeBinary(self, outStr):
        self.writeJSONBase64(outStr)


class TSimpleJSONProtocolFactory:
    def getProtocol(self, trans):
        prot = TSimpleJSONProtocol(trans)
        return prot

File: thrift/lib/hack/src/protocol/simplejson/TSimpleJSONProtocol.php
<?hh // partial

/*
 * Copyright 2006-present Facebook, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Protocol for encoding/decoding simple json
 */
class TSimpleJSONProtocol extends TProtocol {
  const VERSION_1 = 0x80010000;

  private TMemoryBuffer $buffer;
  private Vector<TSimpleJSONProtocolContext> $contexts;

  <<__Rx>>
  public function __construct(TMemoryBuffer $trans) {
    parent::__construct($trans);
    $this->buffer = $trans;
    $this->contexts = Vector {
      new TSimpleJSONProtocolContext($this->buffer),
    };
  }

  protected function pushListWriteContext(): int {
    return $this->pushWriteContext(
      new TSimpleJSONProtocolListContext($this->buffer),
    );
  }

  protected function pushMapWriteContext(): int {
    return $this->pushWriteContext(
      new TSimpleJSONProtocolMapContext($this->buffer),
    );
  }

  protected function pushListReadContext(): void {
    $this->pushReadContext(
      new TSimpleJSONProtocolListContext($this->buffer),
    );
  }

  protected function pushMapReadContext(): void {
    $this->pushReadContext(
      new TSimpleJSONProtocolMapContext($this->buffer),
    );
  }

  protected function pushWriteContext(TSimpleJSONProtocolContext $ctx): int {
    $this->contexts->add($ctx);
    return $ctx->writeStart();
  }

  protected function popWriteContext(): int {
    $ctx = $this->contexts->pop();
    return $ctx->writeEnd();
  }

  protected function pushReadContext(TSimpleJSONProtocolContext $ctx): void {
    $this->contexts->add($ctx);
    $ctx->readStart();
  }

  protected function popReadContext(): void {
    $ctx = $this->contexts->pop();
    $ctx->readEnd();
  }

  protected function getContext(): TSimpleJSONProtocolContext {
    return $this->contexts->at($this->contexts->count() - 1);
  }

  <<__Override>>
  public function writeMessageBegin($name, $type, $seqid): num {
    return
      $this->getContext()->writeSeparator() +
      $this->pushListWriteContext() +
      $this->writeI32(self::VERSION_1) +
      $this->writeString($name) +
      $this->writeI32($type) +
      $this->writeI32($seqid);
  }

  <<__Override>>
  public function writeMessageEnd(): int {
    return $this->popWriteContext();
  }

  <<__Override>>
  public function writeStructBegin($name): int {
    return
      $this->getContext()->writeSeparator() + $this->pushMapWriteContext();
  }

  <<__Override>>
  public function writeStructEnd(): int {
    return $this->popWriteContext();
  }

  <<__Override>>
  public function writeFieldBegin($fieldName, $fieldType, $fieldId): int {
    return $this->writeString($fieldName);
  }

  <<__Override>>
  public function writeFieldEnd(): int {
    return 0;
  }

  <<__Override>>
  public function writeFieldStop(): int {
    return 0;
  }

  <<__Override>>
  public function writeMapBegin($keyType, $valType, $size): int {
    return
      $this->getContext()->writeSeparator() + $this->pushMapWriteContext();
  }

  <<__Override>>
  public function writeMapEnd(): int {
    return $this->popWriteContext();
  }

  <<__Override>>
  public function writeListBegin($elemType, $size): int {
    return
      $this->getContext()->writeSeparator() + $this->pushListWriteContext();
  }

  <<__Override>>
  public function writeListEnd(): int {
    return $this->popWriteContext();
  }

  <<__Override>>
  public function writeSetBegin($elemType, $size): int {
    return
      $this->getContext()->writeSeparator() + $this->pushListWriteContext();
  }

  <<__Override>>
  public function writeSetEnd(): int {
    return $this->popWriteContext();
  }

  <<__Override>>
  public function writeBool($value): int {
    $x = $this->getContext()->writeSeparator();
    if ($value) {
      $this->buffer->write('true');
      $x += 4;
    } else {
      $this->buffer->write('false');
      $x += 5;
    }
    return $x;
  }

  <<__Override>>
  public function writeByte($value): int {
    return $this->writeNum((int) $value);
  }

  <<__Override>>
  public function writeI16($value): int {
    return $this->writeNum((int) $value);
  }

  <<__Override>>
  public function writeI32($value): int {
    return $this->writeNum((int) $value);
  }

  <<__Override>>
  public function writeI64($value): int {
    return $this->writeNum((int) $value);
  }

  <<__Override>>
  public function writeDouble($value): int {
    return $this->writeNum((float) $value);
  }

  <<__Override>>
  public function writeFloat($value): int {
    return $this->writeNum((float) $value);
  }

  protected function writeNum($value): int {
    $ctx = $this->getContext();
    $ret = $ctx->writeSeparator();
    if ($ctx->escapeNum()) {
      $value = (string) $value;
    }

    $enc = PHP\json_encode($value);
    $this->buffer->write($enc);

    return $ret + Str\length($enc);
  }

  <<__Override>>
  public function writeString($value): int {
    $ctx = $this->getContext();
    $ret = $ctx->writeSeparator();
    $value = (string) $value;

    $sb = '"';
    $len = Str\length($value);
    for ($i = 0; $i < $len; $i++) {
      $c = $value[$i];
      $ord = PHP\ord($c);
      switch ($ord) {
        case 8:
          $sb .= '\b';
          break;
        case 9:
          $sb .= '\t';
          break;
        case 10:
          $sb .= '\n';
          break;
        case 12:
          $sb .= '\f';
          break;
        case 13:
          $sb .= '\r';
          break;
        case 34:
          // "
        case 92:
          // \
          $sb .= '\\';
          $sb .= $c;
          break;
        default:
          if ($ord < 32 || $ord > 126) {
            $sb .= '\\u00';
            $sb .= PHP\bin2hex($c);
          } else {
            $sb .= $c;
          }
          break;
      }
    }
    $sb .= '"';
    $this->buffer->write($sb);

    return $ret + Str\length($sb);
  }

  <<__Override>>
  public function readMessageBegin(
    inout $name,
    inout $type,
    inout $seqid,
  ): void {
    throw new TProtocolException(
      'Reading with TSimpleJSONProtocol is not supported. '.
      'Use readFromJSON() on your struct',
    );
  }

  <<__Override>>
  public function readMessageEnd(): void {
    throw new TProtocolException(
      'Reading with TSimpleJSONProtocol is not supported. '.
      'Use readFromJSON() on your struct',
    );
  }

  <<__Override>>
  public function readStructBegin(inout $_name): void {
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $this->pushMapReadContext();
  }

  <<__Override>>
  public function readStructEnd(): void {
    $this->popReadContext();
  }

  <<__Override>>
  public function readFieldBegin(
    inout $name,
    inout $fieldType,
    inout $fieldId,
  ): void {
    $fieldId = null;
    $ctx = $this->getContext();
    $name = null;
    while ($name === null) {
      if ($ctx->readContextOver()) {
        $fieldType = TType::STOP;
        break;
      } else {
        $ctx->readSeparator();
        $this->skipWhitespace();
        $name = $this->readJSONString()[0];
        // We need to guess the type of the value, in case the name is bogus or we are in a skip method up the stack
        $offset = $this->skipWhitespace(true);
        $this->expectChar(':', true, $offset);
        $offset += 1 + $this->skipWhitespace(true, $offset + 1);
        $c = $this->buffer->peek(1, $offset);
        if ($c === 'n' && $this->buffer->peek(4, $offset) === 'null') {
          // We actually want to skip this field, but there isn't an appropriate
          // TType to send back. So instead, we will silently skip
          $ctx->readSeparator();
          $this->skipWhitespace();
          $this->buffer->readAll(4);
          $name = null;
          continue;
        } else {
          $fieldType = $this->guessFieldTypeBasedOnByte($c);
        }
      }
    }
  }

  <<__Override>>
  public function readFieldEnd(): void {
    // Do nothing
  }

  <<__Override>>
  public function readMapBegin(
    inout $keyType,
    inout $valType,
    inout $size,
  ): void {
    $size = null;
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $this->pushMapReadContext();
    if ($this->readMapHasNext()) {
      // We need to guess the type of the keys/values, in case we are in a skip method up the stack
      $keyType = TType::STRING;
      $this
        ->skipWhitespace(); // This is not a peek, since we can do this safely again
      $offset = $this->readJSONString(true)[1];
      $offset += $this->skipWhitespace(true, $offset);
      $this->expectChar(':', true, $offset);
      $offset += 1 + $this->skipWhitespace(true, $offset + 1);
      $c = $this->buffer->peek(1, $offset);
      $valType = $this->guessFieldTypeBasedOnByte($c);
    }
  }

  <<__Override>>
  public function readMapHasNext(): bool {
    return !$this->getContext()->readContextOver();
  }

  <<__Override>>
  public function readMapEnd(): void {
    $this->popReadContext();
  }

  <<__Override>>
  public function readListBegin(inout $elemType, inout $size): void {
    $size = null;
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $this->pushListReadContext();
    if ($this->readListHasNext()) {
      // We need to guess the type of the values, in case we are in a skip method up the stack
      $this->skipWhitespace(); // This is not a peek, since we can do this safely again
      $c = $this->buffer->peek(1);
      $elemType = $this->guessFieldTypeBasedOnByte($c);
    }
  }

  <<__Override>>
  public function readListHasNext(): bool {
    return !$this->getContext()->readContextOver();
  }

  <<__Override>>
  public function readListEnd(): void {
    $this->popReadContext();
  }

  <<__Override>>
  public function readSetBegin(inout $elemType, inout $size): void {
    $size = null;
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $this->pushListReadContext();
    if ($this->readSetHasNext()) {
      // We need to guess the type of the values, in case we are in a skip method up the stack
      $this->skipWhitespace(); // This is not a peek, since we can do this safely again
      $c = $this->buffer->peek(1);
      $elemType = $this->guessFieldTypeBasedOnByte($c);
    }
  }

  <<__Override>>
  public function readSetHasNext(): bool {
    return !$this->getContext()->readContextOver();
  }

  <<__Override>>
  public function readSetEnd(): void {
    $this->popReadContext();
  }

  <<__Override>>
  public function readBool(inout $value): void {
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $c = $this->buffer->readAll(1);
    switch ($c) {
      case 't':
        $value = true;
        $target = 'rue';
        break;
      case 'f':
        $value = false;
        $target = 'alse';
        break;
      case '0':
        $value = false;
        $target = '';
        break;
      case '1':
        $value = true;
        $target = '';
        break;
      default:
        throw new TProtocolException(
          'TSimpleJSONProtocol: Expected t or f, encountered 0x'.
          PHP\bin2hex($c),
        );
    }

    for ($i = 0; $i < Str\length($target); $i++) {
      $this->expectChar($target[$i]);
    }
  }

  protected function readInteger(?int $min, ?int $max): int {
    $val = PHP\intval($this->readNumStr());
    if (($min !== null && $max !== null) && ($val < $min || $val > $max)) {
      throw new TProtocolException(
        'TProtocolException: value '.
        $val.
        ' is outside the expected bounds',
      );
    }
    return $val;
  }

  protected function readNumStr(): string {
    $ctx = $this->getContext();
    if ($ctx->escapeNum()) {
      $this->expectChar('"');
    }
    $count = 0;
    $reading = true;
    while ($reading) {
      $c = $this->buffer->peek(1, $count);
      switch ($c) {
        case '+':
        case '-':
        case '.':
        case '0':
        case '1':
        case '2':
        case '3':
        case '4':
        case '5':
        case '6':
        case '7':
        case '8':
        case '9':
        case 'E':
        case 'e':
          $count++;
          break;
        default:
          $reading = false;
          break;
      }
    }
    $str = $this->buffer->readAll($count);
    if (!PHP\fb\preg_match_simple(
          '/^[+-]?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?$/',
          $str,
        )) {
      throw new TProtocolException(
        'TSimpleJSONProtocol: Invalid json number '.$str,
      );
    }
    if ($ctx->escapeNum()) {
      $this->expectChar('"');
    }

    return $str;
  }

  <<__Override>>
  public function readByte(inout $value): void {
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $value = $this->readInteger(-0x80, 0x7f);
  }

  <<__Override>>
  public function readI16(inout $value): void {
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $value = $this->readInteger(-0x8000, 0x7fff);
  }

  <<__Override>>
  public function readI32(inout $value): void {
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $value = $this->readInteger(-0x80000000, 0x7fffffff);
  }

  <<__Override>>
  public function readI64(inout $value): void {
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $value = $this->readInteger(null, null);
  }

  <<__Override>>
  public function readDouble(inout $value): void {
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $value = (float)$this->readNumStr();
  }

  <<__Override>>
  public function readFloat(inout $value): void {
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $value = (float)$this->readNumStr();
  }

  <<__Override>>
  public function readString(inout $value): void {
    $this->getContext()->readSeparator();
    $this->skipWhitespace();
    $value = $this->readJSONString()[0];
  }

  protected function readJSONString(
    bool $peek = false,
    int $start = 0,
  ): Pair<string, int> {
    if (!$peek) {
      $start = 0;
    }
    $this->expectChar('"', $peek, $start);
    $count = $peek ? 1 : 0;
    $sb = '';
    $reading = true;
    while ($reading) {
      $c = $this->buffer->peek(1, $start + $count);
      switch ($c) {
        case '"':
          $reading = false;
          break;
        case '\\':
          $count++;
          $c = $this->buffer->peek(1, $start + $count);
          switch ($c) {
            case '\\':
              $count++;
              $sb .= '\\';
              break;
            case '"':
              $count++;
              $sb .= '"';
              break;
            case 'b':
              $count++;
              $sb .= PHP\chr(0x08);
              break;
            case '/':
              $count++;
              $sb .= '/';
              break;
            case 'f':
              $count++;
              $sb .= "\f";
              break;
            case 'n':
              $count++;
              $sb .= "\n";
              break;
            case 'r':
              $count++;
              $sb .= "\r";
              break;
            case 't':
              $count++;
              $sb .= "\t";
              break;
            case 'u':
              $count++;
              $this->expectChar('0', true, $start + $count);
              $this->expectChar('0', true, $start + $count + 1);
              $count += 2;
              $sb .= PHP\hex2bin($this->buffer->peek(2, $start + $count));
              $count += 2;
              break;
            default:
              throw new TProtocolException(
                'TSimpleJSONProtocol: Expected Control Character, found 0x'.
                PHP\bin2hex($c),
              );
          }
          break;
        case '':
          // end of buffer, this string is unclosed
          $reading = false;
          break;
        default:
          $count++;
          $sb .= $c;
          break;
      }
    }

    if (!$peek) {
      $this->buffer->readAll($count);
    }

    $this->expectChar('"', $peek, $start + $count);
    return Pair {$sb, $count + 1};
  }

  protected function skipWhitespace(bool $peek = false, int $start = 0): int {
    if (!$peek) {
      $start = 0;
    }
    $count = 0;
    $reading = true;
    while ($reading) {
      $byte = $this->buffer->peek(1, $count + $start);
      switch ($byte) {
        case ' ':
        case "\t":
        case "\n":
        case "\r":
          $count++;
          break;
        default:
          $reading = false;
          break;
      }
    }
    if (!$peek) {
      $this->buffer->readAll($count);
    }

    return $count;
  }

  protected function expectChar(
    string $char,
    bool $peek = false,
    int $start = 0,
  ): void {
    if (!$peek) {
      $start = 0;
    }
    if ($peek) {
      $c = $this->buffer->peek(1, $start);
    } else {
      $c = $this->buffer->readAll(1);
    }

    if ($c !== $char) {
      throw new TProtocolException(
        'TSimpleJSONProtocol: Expected '.
        $char.
        ', but encountered 0x'.
        PHP\bin2hex($c),
      );
    }
  }

  protected function guessFieldTypeBasedOnByte(string $byte): ?int {
    switch ($byte) {
      case '{':
        return TType::STRUCT;
      case '[':
        return TType::LST;
      case 't':
      case 'f':
        return TType::BOOL;
      case '-':
      case '0':
      case '1':
      case '2':
      case '3':
      case '4':
      case '5':
      case '6':
      case '7':
      case '8':
      case '9':
        // These technically aren't allowed to start JSON floats, but are here for backwards compatibility
      case '+':
      case '.':
        return TType::DOUBLE;
      case '"':
        return TType::STRING;
      case ']':
      case '}':
        // We can get here with empty lists/maps, returning a dummy value
        return TType::STOP;
    }

    throw new TProtocolException(
      'TSimpleJSONProtocol: Unable to guess TType for character 0x'.
      PHP\bin2hex($byte),
    );
  }
}
